pkg/logql/log/ip.go

```diff
@@ -0,0 +1,216 @@
+package log
+
+import (
+	"errors"
+	"fmt"
+	"unicode"
+
+	"inet.af/netaddr"
+)
+
+var (
+	errInvalidPattern = errors.New("invalid pattern")
+)
+
+type IPMatchType int
+
+const (
+	IPv4Charset = "0123456789."
+	IPv6Charset = "0123456789abcdefABCDEF:."
+
+	IPMatchLine IPMatchType = iota
+	IPMatchLabel
+)
+
+// Should be one of the netaddr.IP, netaddr.IPRange, netadd.IPPrefix.
+type IPMatcher interface{}
+
+type IPFilter struct {
+	pattern   string
+	matcher   IPMatcher
+	matchType IPMatchType
+
+	// if used as label matcher, this holds the identifier label name.
+	// e.g: (|remote_addr = ip("xxx")). Here labelName is `remote_addr`
+	labelName string
+
+	// patternError represents any invalid pattern provided to match the ip.
+	patternError error
+}
+
+// IPFilter search for IP addresses of given `pattern` in the given `line`.
+// It returns true if pattern is matched with at least one IP in the `line`
+
+// pattern - can be of the following form for both IPv4 and IPv6.
+// 1. SINGLE-IP - "192.168.0.1"
+// 2. IP RANGE  - "192.168.0.1-192.168.0.23"
+// 3. CIDR      - "192.168.0.0/16"
+func NewIPFilter(pattern string) *IPFilter {
+	filter := &IPFilter{pattern: pattern}
+
+	matcher, err := getMatcher(pattern)
+	filter.matcher = matcher
+	filter.patternError = err
+
+	return filter
+}
+
+func NewIPLabelFilter(pattern string, label string) *IPFilter {
+	filter := NewIPFilter(pattern)
+	filter.labelName = label
+	filter.matchType = IPMatchLabel
+	return filter
+}
+
+func (ipf *IPFilter) Filter(line string) bool {
+	if len(line) == 0 {
+		return false
+	}
+
+	n := len(line)
+
+	filterFn := func(line string, start int, charset string) (bool, int) {
+		iplen := stringSpan(line[start:], charset)
+		if iplen < 0 {
+			return false, 0
+		}
+		ip, err := netaddr.ParseIP(line[start : start+iplen])
+		if err == nil {
+			if contains(ipf.matcher, ip) {
+				return true, 0
+			}
+		}
+		return false, iplen
+	}
+
+	// This loop try to extract IPv4 or IPv6 address from the arbitrary string.
+	// It uses IPv4 and IPv6 prefix hints to find the IP addresses faster without using regexp.
+	for i := 0; i < n; i++ {
+		if i+3 < n && ipv4Hint([4]byte{line[i], line[i+1], line[i+2], line[i+3]}) {
+			ok, iplen := filterFn(line, i, IPv4Charset)
+			if ok {
+				return true
+			}
+			i += iplen
+			continue
+		}
+
+		if i+4 < n && ipv6Hint([5]byte{line[i], line[i+1], line[i+2], line[i+3], line[i+4]}) {
+			ok, iplen := filterFn(line, i, IPv6Charset)
+			if ok {
+				return true
+			}
+			i += iplen
+			continue
+		}
+	}
+	return false
+}
+
+// `Process` implements `Stage` interface
+func (ipf *IPFilter) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+
+	// make sure the pattern provided was valid, even before trying to match.
+	if ipf.patternError != nil {
```

Comment:
This can be spotted before and so should return an error during query parsing.

Reply:
Yea. thought about that actually. My previous implementation had thrown this error during constructor of `NewIPFilter()` but while writing grammar, there was no way to handle that error while creating the `LabelFilter` expression!. So I changed it to like this. 

May be I'm missing something about handling this error during the parse stage?. Can you point me any example?

Reply:
You can use the `Stage()` method of this expr, it can return an errors and will be considered as 400.